src/lib/formatter/SarifFormatter.ts

```diff
@@ -0,0 +1,247 @@
+import { deepCopy } from '../util/Utils';
+import { ENGINE, PMD_VERSION } from '../../Constants';
+import { Rule, RuleResult, RuleViolation } from '../../types';
+import * as url from 'url';
+import { RuleCatalog } from '../services/RuleCatalog';
+import { Controller } from '../../Controller';
+import { CLIEngine } from 'eslint';
+import * as retire from 'retire';
+
+const ERROR = 'error';
+const WARNING = 'warning';
+
+/**
+ * Formatter based on https://docs.oasis-open.org/sarif/sarif/v2.0/csprd02/sarif-v2.0-csprd02.html
+ */
+abstract class SarifFormatter {
```

Comment:
Could we get some more documentation on the methods and classes in this file?

Reply:
@jfeingold35 I have pushed another commit with better docs.